design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+}
+
+@range(min: 1)
+integer PositiveInteger
+```
+
+In the above example,
+
+> the range trait applied to numberOfItems takes precedence over the one
+> applied to PositiveInteger. The resolved minimum will be 7, and the maximum
+> 12.
+
+When the constraint trait is applied to a member shape, the tuple struct's name
+will be the PascalCased name of the member shape, `NumberOfItems`.
+
+[the spec]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html?highlight=enum#trait-precedence
+
+Unresolved questions
+--------------------
+
+1. Should we code-generate unsigned integer types (`u16`, `u32`, `u64`) when
+   the `range` trait is applied with `min` set to a value greater than or equal
+   to 0?
+    - A user has even suggested to use the `std::num::NonZeroUX` types (e.g.
+      [`NonZeroU64`]) when `range` is applied with `min` set to a value greater
+      than 0.
+2. In request deserialization, should we fail with the first violation and
+   immediately render a response, or attempt to parse the entire request and
+   provide a complete and structured report?
+3. Should we provide a mechanism for the service implementer to construct a
+   Rust type violating the modeled constraints in their business logic e.g. a
+   `T::new_unchecked()` constructor? This could be useful (1) when the user
+   _knows_ the provided inner value does not violate the constraints and
+   doesn't want to incur the performance penalty of the check; (2) when the
+   struct is in a transient invalid state. However:
+    - (2) is arguably a modelling mistake and a separate struct to represent
+      the transient state would be a better approach,
+    - the user can always use `unsafe` Rust to bypass the validation; and
+    - adding this constructor is a backwards-compatible change, so it can
+      always be added later if this feature is requested.
+
+[`NonZeroU64`]: https://doc.rust-lang.org/std/num/struct.NonZeroU64.html
+
+Alternative design
+------------------
+
+An alternative design with less public API surface would be to perform
+constraint validation at request deserialization, but hand over a regular
+"loose" type (e.g. `String` instead of `NiceString`) that allows for values
+violating the constraints. If we were to implement this approach, we can
+implement it by wrapping the incoming value in the aforementioned tuple struct
+to perform the validation, and immediately unwrap it.
+
+Comparative advantages:
+
+* Validation remains an internal detail of the framework. If the semantics of a
+  constraint trait change, the behavior of the service is still
+  backwards-incompatibly affected, but user code is not.
+* Less "invasive". Baking validation in the generated type might be deemed as
+  the service framework overreaching responsibilities.
+
+Comparative disadvantages:
+
+* It becomes _possible_ to send responses with invalid operation outputs. All
+  the service framework could do is log the validation errors.
```

Comment:
Should we really care about this? At the end the server is the authority and we do not validate outputs..

Reply:
At the very least I'd consider it to likely be an implementation error from the user with regards to what they modeled, and as such performing validation and logging warnings would help them identify it. So we _should_ care, but very little.

Reply:
As a service owner you probably do care that your outputs do not match your model: your server is breaking the contract you want to set with your clients.
So being able to have clear metrics about it seems important. Consistent logging would allow that.